### PR TITLE
Basics of Cert-Count Non-Locking Telemetry

### PR DESCRIPTION
**Pull request by kitography**
This is the basics of the Certificate Count Telemetry.

Lock is Removed.  See Comment.
- [x] Test sort and duplicate-remove
- [x] Fix "early" reporting of count metric.
- [x] Add count-metric to tidy-status (with warning about search)

When we store a cert now, if certs haven't been counted, that is added to a slice which will be checked for double counting.

I'm pretty sure I got every place that a cert is stored:

```kit@gelfly pki % git grep -n "\.Put("
cert_util.go:218:       if err = req.Storage.Put(ctx, certEntry); err != nil {
* This is path-update for a legacy cert storage.  This could plausibly result in a double counting during initialise (if the 'read' - which happens first - is picked up, but not the 'delete').  Unfortunately, I'm not sure there's anything to be done about that without a lot more complexity (and considerations of error cases).
cert_util_test.go:42:           err := storage.Put(context.Background(), &logical.StorageEntry{
* This is a test
cert_util_test.go:77:           err := storage.Put(context.Background(), &logical.StorageEntry{
* This is also a test
crl_util.go:235:                err = req.Storage.Put(ctx, revEntry)
* Increments when we successfully write a revoked cert
crl_util.go:567:                        err = req.Storage.Put(ctx, revokedEntry)
* This is an update call, not a new entry
crl_util.go:643:        err = sc.Storage.Put(sc.Context, &logical.StorageEntry{
* Path here is crls/ (not a cert)
path_config_crl.go:112: err = req.Storage.Put(ctx, entry)
* Path here is config/crl (not a cert)
path_config_urls.go:92: err = storage.Put(ctx, entry)
* Path here is urls/ (not a cert)
path_issue_sign.go:413:         err = req.Storage.Put(ctx, &logical.StorageEntry{
* We increment the certificate count here
path_roles.go:608:              if err := s.Put(ctx, jsonEntry); err != nil {
* Path here is role/ (not a cert)
path_roles.go:762:      if err := req.Storage.Put(ctx, jsonEntry); err != nil {
* Path here is role/ (not a cert)
path_roles.go:967:      if err := req.Storage.Put(ctx, jsonEntry); err != nil {
* Path here is role/ (not a cert)
path_roles_test.go:79:  if err := storage.Put(context.Background(), entry); err != nil {
* Path here is role/testrole (not a cert)
path_roles_test.go:179: if err := storage.Put(context.Background(), entry); err != nil {
* Path here is role/testrole (not a cert)
path_roles_test.go:275: if err := storage.Put(context.Background(), entry); err != nil {
* Path here is role/testrole (not a cert)
path_roles_test.go:373: if err := storage.Put(context.Background(), entry); err != nil {
* Path here is role/testrole (not a cert)
path_root.go:260:       err = req.Storage.Put(ctx, &logical.StorageEntry{
* We increment after a successful save
path_root.go:444:       err = req.Storage.Put(ctx, &logical.StorageEntry{
* We increment after a successful save
storage.go:222: return sc.Storage.Put(sc.Context, json)
* This is keys, not certs
storage.go:566: return sc.Storage.Put(sc.Context, json)
* This is issuer, not certs
storage.go:739: return sc.Storage.Put(sc.Context, json)
* This is CRL Config, not certs
storage.go:772: return sc.Storage.Put(sc.Context, json)
* This is keys config, not certs
storage.go:797: return sc.Storage.Put(sc.Context, json)
* This is issuers config, not certs
storage_migrations.go:167:      return s.Put(ctx, json)
* This is legacy bundle, not certs
storage_migrations_test.go:77:  err = s.Put(ctx, json)
* Also legacy bundle, not certs
storage_migrations_test.go:182: err = s.Put(ctx, json)
* Also legacy bundle, not certs
```

### CODE REVIEW

File: builtin/logical/pki/path_root.go
Comment (kitography) [inline builtin/logical/pki/path_root.go:265]: My thinking here (on not using a defer) is that CRL Building can be pretty long, but if others feel like a defer is still better, I'd change it back.

File: builtin/logical/pki/path_tidy.go
Comment (stevendpclark) [inline builtin/logical/pki/path_tidy.go:374]: I'm a bit puzzled by these locking calls here and above in `tidyStatusIncCertStoreCount`? We aren't updating storage, or we have already, and the atomic operations should cover us for the inc/dec of the counter?

File: builtin/logical/pki/path_root.go
Comment (cipherboy) [reply builtin/logical/pki/path_root.go:265]: I think CRL building can be _reallly_ long so I like not holding this lock (if we keep this lock).

File: builtin/logical/pki/path_root.go
Comment (cipherboy) [inline builtin/logical/pki/path_root.go:269]: I'm not really inclined to increment here for generated roots, tbh. While in the storage, they're mostly exposed there for $LEGACY_REASONS aren't traditional certs that you'd want to cleanup (until they expire...). Plus, they're duplicated into Issuers storage which is the real "canonical" source if you would.

Comment (kitography) [reply builtin/logical/pki/path_root.go]: It doesn't look like, other than reading/parsing the cert, there's an easy way to tell whether a stored cert is a root or leaf.  So, if we removed this, I'd be worried that someone will see 'cool no certs', then turn the node on-and-off 'where did these nine certs come from!?' (when the list count-completes).

Checking: it looks like these currently appear in tidy-counts too, so if we removed them, the tidy-count could be higher than the certs-in-storage too.  I think that would be more confusing.

Comment (cipherboy) [reply builtin/logical/pki/path_root.go]: Ahhhh yeah, good points :-) I'll retract this suggestion then. :D

**Comment by kitography** [comment]
For anyone looking in: after discussion internally, we've settled on it being more important not to lock (and block all issuance and revocation) on startup than to have perfectly correct metrics here.  Instead, we'll do a "best effort" metric - which might include some double counting of certificate entries.

If we do miscount, we want to make sure we 'overcount' rather than undercount the number of certs in storage.

( https://hashicorp.slack.com/archives/C0386B7KPHR/p1660332542341629 )

File: builtin/logical/pki/path_tidy.go
Comment (kitography) [reply builtin/logical/pki/path_tidy.go]: This lock was basically to avoid writing to the certstore, while the initial count (list operation) was being performed.

Resolving this conversation, since we've decided no more lock.

File: builtin/logical/pki/path_root.go
Comment (kitography) [reply builtin/logical/pki/path_root.go]: Thank you - but this lock is removed.  :(

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [inline builtin/logical/pki/backend.go:422]: Would it not be better to move both calls to SetGauge till after the for loops with the appropriate values so they only appear on the metrics with the "correct value"?

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [inline builtin/logical/pki/backend.go:506]: Should we assign nil to `b.possibleDoubleCountedSerials` and `b.possibleDoubleCountedRevokedSerials` to release the memory after we are done?

Also did I miss it or did we not add the values to the counter from these lists?

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [inline builtin/logical/pki/backend.go:205]: This would be better if it were an atomic boolean with how we are using it across threads at the moment.

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [inline builtin/logical/pki/backend.go:525]: I'm of two minds here, part of me really wants to get rid of the `certsCounted` boolean parameter here and just embed it within the case as `b.certsCounted` so callers don't have to worry/get it wrong about where this boolean should be coming from. 

I see the point of grabbing it before/after storage calls and passing it in but I support my fear of someone using this badly passing in a false value leading to a mem leak is making me nervous.

**Review by stevendpclark — COMMENTED** [review]
Good stuff, I do like where this is headed. Sorry though I went a little nuts on the comments...

File: builtin/logical/pki/backend.go
Comment (kitography) [reply builtin/logical/pki/backend.go]: Yeah, that's fair, I've updated to include this.

File: builtin/logical/pki/backend.go
Comment (kitography) [reply builtin/logical/pki/backend.go:424]: I'll let others weigh in here, but wanted to outline my thinking:
* This would only ever be "high" (a 'correct' value for our contract)
* "Fixing" this would require far more code complexity (another backend-bool measure, or another set of decrement calls), since the decrement calls below also report metrics.
* Sorting long lists could take a very, very long time, in precisely the case we would want these metrics "immediately"

File: builtin/logical/pki/backend.go
Comment (kitography) [reply builtin/logical/pki/backend.go:461]: I've moved the sort of possibleDoubleCountedRevokedSerials after this.

I'm not clear how moving the sorting of revokedEntries later here would help. My thinking is :
* we want as much time between setting the certsCounted boolean, and sorting the two dynamic lists as possible.
* revokedEntries and entries aren't being modified (after the initial list operation where they are set), so sorting them is safe (and might take a while).
* so let's do the safe-slow thing first to buy time.

File: builtin/logical/pki/backend.go
Comment (kitography) [reply builtin/logical/pki/backend.go:535]: Walking the middle line:
* I've added a comment on how to use certsCounted.
* I've added a check that the list to append to isn't nil.  This should prevent memory leak.
- I don't know how to make this nil-check atomic, so if you have ideas, or a willing to chat about it, that would be really nice.  I'm hesitant to add a recover() from a panic here, since like, oom should panic.

File: builtin/logical/pki/backend.go
Comment (kitography) [reply builtin/logical/pki/backend.go]: Fixing this - I've add it to the tidy-status call with a warning that deduplication isn't complete.

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [reply builtin/logical/pki/backend.go:492]: Yup your thinking makes perfect sense, thanks for the update.

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [reply builtin/logical/pki/backend.go:565]: Yeah this is a good middle ground, as we've discussed I believe this week, I think we are okay with the nil-check not being 100% atomic here and we shouldn't have any issues with a panic here as go will create a new array under the hood within the append call. Please correct me if I'm incorrect here!

File: builtin/logical/pki/backend.go
Comment (stevendpclark) [inline builtin/logical/pki/backend.go:577]: I'm pretty sure this can't happen in the current code/usage I'm seeing with the two decrements, but should we add a guard here for b.revokedCertCount being 0 and we under-flowing the unit32? The only safe way I can think of this is this ugly code block

```go
testInt := uint32(1)

for i := 0; i <= 5; i++ {
	oldVal := atomic.LoadUint32(&testInt)
	if oldVal == 0 {
		break
	}
	newVal := oldVal - 1
	if atomic.CompareAndSwapUint32(&testInt, oldVal, newVal) {
		break
	}
}
```

File: builtin/logical/pki/cert_util.go
Comment (stevendpclark) [inline builtin/logical/pki/cert_util.go:237]: Sorry I missed this in the original review, but I believe we need to be passing in just the serial numbers, not the entire paths, within the `incrementTotalRevokedCertificatesCount` and `incrementTotalCertificatesCount` calls. If I understand correctly we will never match as the list operation for say "/certs" only returns the key name not the entire path and key above in the `initializeStoredCertificateCounts` function.

Comment (stevendpclark) [reply builtin/logical/pki/cert_util.go:237]: Or stripping out the `revoked` and/or `certs` path prefix within `incrementTotalRevokedCertificatesCount` and `incrementTotalCertificatesCount`

Comment (kitography) [reply builtin/logical/pki/cert_util.go:237]: We will match (the test passes), because of the slice indexing (line 611 of backend.go) right now.  I can see why stripping out the `revoked` and or `certs` prefix in each function might be cleaner, though it would also be a bit slower.

Comment (stevendpclark) [reply builtin/logical/pki/cert_util.go:237]: ah yes oops I missed that slice operation, thanks for pointing it out! 

Perhaps we could make the slicing operation a little more obvious by extracting it out into a var that is reused across the three tests within the for loop, and not across the code base?

Comment (kitography) [reply builtin/logical/pki/cert_util.go:237]: Okay, I've moved this to the helper function (the other side), slightly less ugly, and it doesn't seem to be performing any slower when I test it.  Does that work?

File: builtin/logical/pki/backend.go
Comment (kitography) [reply builtin/logical/pki/backend.go]: I admit I don't like this.  If someone ever saw an underflow it would be solvable by "turning off and turning on" and would be a really really helpful report that we aren't honouring the promise of "always overcount".

If someone else agrees with this solution, I'll go ahead and add it, but in general I don't think we can protect against people writing bad code in the future with uglier code now.

File: builtin/logical/pki/cert_util.go
Comment (stevendpclark) [reply builtin/logical/pki/cert_util.go:237]: Yup, even better than what I was asking, thank you!

**Review by stevendpclark — APPROVED** [review]
Looks good to me, thanks for addressing all my concerns overall.

**Pull request merged**
